Separate service id from task

pbetkier opened the pull request:
Consul service used to be registered under the same id as its related marathon task id. This PR separates these two ids. Marathon task id is remembered in service's `marathon-task` tag. Also did some refactorings: extracted Service and ServiceId domain types, and also ServiceRegistry interface to make things more clear by hiding Consul internals.

I'm still working on this PR, going to cleanup some code and verify that all branches are covered with tests. But the main idea is there, so waiting for your feedback.

It's a prerequisite to #89, because we would register more than one service for a given marathon task.

janisz commented:
I think this should be split into 2 PR: introduce Service and ServiceId domain types and change ServiceId to generated uuid. This will make review and testing easier. Another thing changing IDs could be harmful so we need to test impact of this action.

pbetkier commented:
Even if do we split, we have to test this PR thoroughly because it's a significant change. Why not do it once? I've split the change into two separate commits for the review's sake, but I don't see the point in splitting it into separate PRs. One would contain the feature, but introducing complexity in the codebase and the other would only clean things up. Still, we merge either both or nothing IMO :)

janisz commented:
I thought we can merge refactoring first and then ID change. With refactoring most of tests should remain untouched and by this we will ensure change doesn't break anything.

janisz commented:
I created dedicated task (#100) for extracting domain service object. It should be backward comaptible and we could release it as 4.1 build with Go 1.7.

janisz commented on service/service.go:
```diff
@@ -0,0 +1,44 @@
+	"strings"
+
+	"github.com/allegro/marathon-consul/apps"
+	"fmt"
```
`find ./ -name '*.go' | xargs goimports -w`

janisz commented on service/service.go:
```diff
@@ -0,0 +1,44 @@
+	RegisteringAgentAddress string
+}
+
+func (s *Service) TaskId() (apps.TaskId, error) {
```
Why pointer receiver?

janisz commented on consul/consul.go:
```diff
@@ -63,58 +56,71 @@ func (c *Consul) getServicesUsingProviderWithRetriesOnAgentFailure(provide Servi
 	return allInstances, nil
 }
 
+func (c *Consul) consulServiceToService(consulService *consulapi.CatalogService) *service.Service {
```
How about moving this function to `service` package and make it constructor function e.g. `func FromConsul(consulService *consulapi.CatalogService) *service.Service`

pbetkier replied:
Then service package would depend on consul package, not cool :)

janisz replied:
OK. But this could be a pure function not depended on `*Consul`

pbetkier replied:
Good point, I'll make it a pure function.

janisz commented on consul/consul.go:
```diff
@@ -1,32 +1,25 @@
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/allegro/marathon-consul/apps"
+	"github.com/allegro/marathon-consul/service"
```
Consul package shouldn't have dependency to service.

pbetkier replied:
I'd say right the opposite, you don't want our domain types with business logic depend on infrastructure, external classes like consul.

janisz replied:
Good point

janisz commented on consul/consul.go:
```diff
@@ -162,9 +168,32 @@ func (c *Consul) register(service *consulapi.AgentServiceRegistration) error {
 }
 
-func (c *Consul) Deregister(serviceId apps.TaskId, agentAddress string) error {
+func (c *Consul) DeregisterByTask(taskToDeregister apps.TaskId, agentAddress string) error {
```
`taskToDeregister` → `taskIDToDeregister` or just `taskID`

medzin commented on consul/consul.go:
```diff
@@ -35,24 +28,24 @@ func New(config ConsulConfig) *Consul {
 		if services, err := provide(agent.Client); err != nil {
 			log.WithError(err).WithField("Address", agent.IP).
-				Error("An error occurred getting services from Consul, retrying with another agent")
+			Error("An error occurred getting services from Consul, retrying with another agent")
```
Are you sure it is Error and not a Debug/Notice message?

pbetkier replied:
@janisz what do you think? Warn / Debug? Should we change levels in this PR?

janisz replied:
I'll leave it as is (without changing indentation)

medzin commented on consul/consul.go:
```diff
@@ -162,9 +168,32 @@ func (c *Consul) register(service *consulapi.AgentServiceRegistration) error {
+		return err
+	}
+	c.Deregister(service)
+	return err
```
IMHO - better to be explicit here so`return nil`

pbetkier commented:
@janisz @dankraw I fixed according to review comments, can I merge this into develop?

dankraw commented:
late but 👍  :>
